src/docker-extract.c

```diff
@@ -255,6 +256,21 @@ int extract_tar(const char *tarfile, const char *rootfs_dir) {
             continue;
         }
 
+        // Issue 977 - Force write perms needed for user builds
+        if(getuid() != 0) {
+          #if ARCHIVE_VERSION_NUMBER <= 3000000
+            perms = archive_entry_mode(entry);
+            if( (perms & S_IWUSR) != S_IWUSR) {
+               archive_entry_set_mode(entry, perms | S_IWUSR);
+            }
+          #else
+            perms = archive_entry_perm(entry);
+            if( (perms & S_IWUSR) != S_IWUSR) {
+               archive_entry_set_perm(entry, perms | S_IWUSR);
+            }
+          #endif
+        }
```

Comment:
Could this be refactored to remove some duplicate lines like this:
```
perms = archive_entry_mode(entry);
if( (perms & S_IWUSR) != S_IWUSR) {
    #if ARCHIVE_VERSION_NUMBER <= 3000000
    	archive_entry_set_mode(entry, perms | S_IWUSR);
    #else
    	archive_entry_set_perm(entry, perms | S_IWUSR);
    #endif
}
```
Didn't test, so that might not work, but you get the idea.

And also, could we get 4 space block indents instead of 2 for consistency with the surrounding code?

Reply:
Sorry about the indents... not sure how I ended up with 2 there.

r.e. refactoring, it'd then have 2 separate #if #endif blocks, like this:

```
#if ARCHIVE_VERSION_NUMBER <= 3000000
    perms = archive_entry_mode(entry);
#else
    perms = archive_entry_perm(entry);
#endif
if( (perms & S_IWUSR) != S_IWUSR) {
    #if ARCHIVE_VERSION_NUMBER <= 3000000
    	archive_entry_set_mode(entry, perms | S_IWUSR);
    #else
    	archive_entry_set_perm(entry, perms | S_IWUSR);
    #endif
}
```

I thought the single #if #else block was easier to read, but don't mind - what's your preference :-)?

Reply:
Your absolutely correct.  I missed that detail on my first read.  

OK.  I'm good.  I think @yhcote or another of our more experienced C coders should read it too.